packages/fuels-programs/src/calls/tx_dependency_extension.rs

```diff
@@ -0,0 +1,72 @@
+use fuel_tx::Receipt;
+use fuels_core::types::{
+    bech32::Bech32ContractId,
+    errors::{transaction::Reason, Error, Result},
+};
+
+use crate::calls::utils::{find_id_of_missing_contract, is_missing_output_variables, sealed};
+
+/// How many times to attempt to resolve missing tx dependencies.
+pub const DEFAULT_TX_DEP_ESTIMATION_ATTEMPTS: u64 = 10;
+
+#[async_trait::async_trait]
+pub trait TxDependencyExtension: Sized + sealed::Sealed {
+    async fn simulate(&mut self) -> Result<()>;
```

Comment:
I'm not too happy with it sharing a name with the methods on `CallHandler`s themselves. Fine for normal contract and script calls but for contract multi call it is named the same as the method `simulate` but doesn't do the same thing.

I get this is needed for something akin to the template method pattern but maybe there is a better way.

My playing around produced two alternatives:
1) Move everything that exists only to support `estimate_tx_dependencies` into a `pub(crate)` trait and make it a supertrait of `TxDependencyExtension`. You'll get a warning that the new trait is more private than `TxDependencyExtensions`, you silence this warning. Downside is now users cannot extend `TxDependencyExtensions`, upside is they never could because of `sealed::Sealed`. Not sure if any other downsides exist. Maybe something around using it in a trait bound.
2) A bit more complicated but works around the private trait issue:
Again you create a new trait with everything secondary, for example:
```rust
#[async_trait::async_trait]
pub(crate) trait PrivatePart {
    async fn simulate(&mut self) -> Result<()>;
    /// Appends additional external contracts as dependencies to this call.
    /// Effectively, this will be used to create additional
    /// [`fuel_tx::Input::Contract`]/[`fuel_tx::Output::Contract`]
    /// pairs and set them into the transaction. Note that this is a builder
    /// method, i.e. use it as a chain:
    ///
    /// ```ignore
    /// my_contract_instance.my_method(...).append_contract(additional_contract_id).call()
    /// my_script_instance.main(...).append_contract(additional_contract_id).call()
    /// ```
    ///
    /// [`Input::Contract`]: fuel_tx::Input::Contract
    /// [`Output::Contract`]: fuel_tx::Output::Contract
    fn append_contract(&mut self, contract_id: Bech32ContractId);
    /// Appends `num` [`fuel_tx::Output::Variable`]s to the transaction.
    /// Note that this is a builder method, i.e. use it as a chain:
    ///
    /// ```ignore
    /// my_contract_instance.my_method(...).append_variable_outputs(num).call()
    /// my_script_instance.main(...).append_variable_outputs(num).call()
    /// ```
    ///
    /// [`Output::Variable`]: fuel_tx::Output::Variable
    fn append_variable_outputs(&mut self, num: u64);

    fn append_missing_dependencies(&mut self, receipts: &[Receipt]) {
        if is_missing_output_variables(receipts) {
            self.append_variable_outputs(1);
        }
        if let Some(contract_id) = find_id_of_missing_contract(receipts) {
            self.append_contract(contract_id);
        }
    }
}
```
You implement it for all manners of `CallHandlers`:
```rust
#[async_trait::async_trait]
impl<A, C, T> PrivatePart for CallHandler<A, C, T>
where
    A: Account,
    C: Extendable + Buildable + Parsable + Send + Sync,
    T: Tokenizable + Parameterize + Debug + Send + Sync,
{
    async fn simulate(&mut self) -> Result<()> {
        self.simulate().await?;

        Ok(())
    }

    fn append_variable_outputs(&mut self, num: u64) {
        self.call.append_variable_outputs(num);
    }

    fn append_contract(&mut self, contract_id: Bech32ContractId) {
        self.call.append_contract(contract_id);
    }
}

#[async_trait::async_trait]
impl<A> PrivatePart for CallHandler<A, Vec<ContractCall>, ()>
where
    A: Account,
{
    async fn simulate(&mut self) -> Result<()> {
        self.simulate_without_decode().await?;

        Ok(())
    }

    fn append_variable_outputs(&mut self, num: u64) {
        self.call
            .iter_mut()
            .take(1)
            .for_each(|call| call.append_variable_outputs(num));
    }

    fn append_contract(&mut self, contract_id: Bech32ContractId) {
        self.call
            .iter_mut()
            .take(1)
            .for_each(|call| call.append_contract(contract_id.clone()));
    }
}
```
That covers scripts, contract calls, and contract multi calls.

Now you have a private helper:
```rust
struct Helper<Handler> {
    handler: Handler,
}

impl<Handler> Helper<Handler> {
    pub fn new(handler: Handler) -> Self {
        Self { handler }
    }
}
```

Now all `CallHandlers` that implement the private trait `PrivateParts` get a public method:
```rust
#[async_trait::async_trait]
impl<A, C, T> TxDependencyExtension for CallHandler<A, C, T>
where
    CallHandler<A, C, T>: PrivatePart,
    A: Send,
    T: Send,
    C: Send,
{
    async fn estimate_tx_dependencies(self, max_attempts: Option<u64>) -> Result<Self> {
        Helper::new(self)
            .estimate_tx_dependencies(max_attempts)
            .await
    }
}
```
or even simpler
```rust
#[async_trait::async_trait]
impl<T> TxDependencyExtension for T
where
    T: PrivatePart + Send + sealed::Sealed,
{
    async fn estimate_tx_dependencies(self, max_attempts: Option<u64>) -> Result<Self> {
        Helper::new(self)
            .estimate_tx_dependencies(max_attempts)
            .await
    }
}
```

Reply:
After merging the `variable output policy PR`, I think I found an even better option. Do not use a default implementation for `determine_missing_contracts` but implement it for `CallHandler` (contract and script call) and `CallHandler<Vec<ContractCall>>` separately. This means that we do not need the `simulate` method at all in the trait.

Reply:
I think I found an even better solution. I deleted the trait :joy: